File: packages/sb-uhq/src/self/commands/gestion/prefix.js
"use strict";

const fs   = require("fs");
const path = require("path");
const { dataPath } = require("../../func/data-path");

const PREFIX_FILE    = dataPath("config", "prefix.json");
const DEFAULT_PREFIX = ".";

let cachedPrefix = null;

// ── I/O ──────────────────────────────────────────────────────────────────────

function loadPrefix() {
  if (cachedPrefix) return cachedPrefix;
  try {
    if (!fs.existsSync(PREFIX_FILE)) return _save(DEFAULT_PREFIX);
    const data = JSON.parse(fs.readFileSync(PREFIX_FILE, "utf8"));
    cachedPrefix = data.prefix || DEFAULT_PREFIX;
    return cachedPrefix;
  } catch {
    return _save(DEFAULT_PREFIX);
  }
}

function _save(prefix) {
  cachedPrefix = prefix;
  fs.mkdirSync(path.dirname(PREFIX_FILE), { recursive: true });
  fs.writeFileSync(PREFIX_FILE, JSON.stringify({ prefix }, null, 2));
  return prefix;
}

// ── Logique pure (bridge uniquement) ─────────────────────────────────────────

/**
 * @param {import("discord.js-selfbot-v13").Client} _client
 * @param {{ action: "set"|"get", prefix?: string }} payload
 */
async function execute(_client, payload) {
  const { action, prefix: newPrefix } = payload;

  if (action === "get") return { prefix: loadPrefix() };

  if (action === "set") {
    if (!newPrefix || newPrefix.length > 3)
      throw new Error("Le préfixe doit faire entre 1 et 3 caractères.");
    _save(newPrefix);
    return { prefix: newPrefix };
  }

  throw new Error(`Action prefix inconnue : '${action}'`);
}

module.exports = { name: "prefix", execute, loadPrefix };

File: packages/sb-uhq/app/commands/gestion/prefix.py
"""prefix — préfixe des commandes texte du selfbot (bridge + lecture)."""

from __future__ import annotations

from ...func.data_path import data_path, read_json, write_json

PREFIX_FILE = data_path("config", "prefix.json")
DEFAULT_PREFIX = "."

_cached_prefix: str | None = None


def _save(prefix: str) -> str:
    global _cached_prefix
    _cached_prefix = prefix
    write_json(PREFIX_FILE, {"prefix": prefix})
    return prefix


def load_prefix() -> str:
    global _cached_prefix
    if _cached_prefix:
        return _cached_prefix
    if not PREFIX_FILE.exists():
        return _save(DEFAULT_PREFIX)
    data = read_json(PREFIX_FILE, None)
    if not data:
        return _save(DEFAULT_PREFIX)
    _cached_prefix = data.get("prefix") or DEFAULT_PREFIX
    return _cached_prefix


async def execute(client, payload):
    action = payload.get("action")
    if action == "get":
        return {"prefix": load_prefix()}
    if action == "set":
        new_prefix = payload.get("prefix")
        if not new_prefix or len(new_prefix) > 3:
            raise ValueError("Le préfixe doit faire entre 1 et 3 caractères.")
        _save(new_prefix)
        return {"prefix": new_prefix}
    raise ValueError(f"Action prefix inconnue : '{action}'")
